### PR TITLE
custom function edge delivery integration

### PR DESCRIPTION
**Pull request by deepprakash345**

### CODE REVIEW

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/_cq_dialog/.content.xml [hunk 1]
```diff
@@ -71,12 +71,6 @@
                                       fieldDescription="This client library contains JavaScript Functions which can be used in expressions supported by Adaptive Forms."
                                       fieldLabel="Client library category"
                                       name="./clientLibRef"/>
-                                    <specVersion
```
Comment (rismehta) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/_cq_dialog/.content.xml:74]: Why is this removed ?

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/container.html [hunk 2]
```diff
@@ -29,7 +29,10 @@
 </div>
 <sly data-sly-test="${!isEdgeDeliveryRequest}">
 <sly data-sly-test="${wcmmode.edit}" data-sly-use.allowed="com.day.cq.wcm.foundation.AllowedComponentList"></sly>
-<form data-sly-use.grid="com.day.cq.wcm.foundation.model.responsivegrid.ResponsiveGrid"
+
+<form data-sly-use.container="com.adobe.cq.forms.core.components.models.form.FormContainer"
```
Comment (rismehta) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/container.html:33]: form container is already available, why are you instantiating it twice.

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/form/FormConfigurationProviderImpl.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            Resource configResource = configurationResourceResolver.getResource(resource, CUSTOM_FUNCTION_CONFIG_BUCKET_NAME,
+                CUSTOM_FUNCTION_CONFIG_NAME);
+            if (configResource != null) {
+                Resource jcrResource = configResource.getChild(JcrConstants.JCR_CONTENT);
+                if (jcrResource != null) {
+                    ValueMap configValueMap = jcrResource.getValueMap();
+                    String owner = configValueMap.getOrDefault("owner", "").toString();
+                    String repo = configValueMap.getOrDefault("repo", "").toString();
+                    if (!owner.isEmpty() && !repo.isEmpty()) {
+                        customFunctionUrl = "https://main--" + repo + "--" + owner + ".hlx.live/blocks/form/functions.js";
```
Comment (rismehta) [inline bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/form/FormConfigurationProviderImpl.java:59]: Please create constants for this

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/form/FormConfigurationProviderImpl.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+import com.adobe.cq.forms.core.components.models.form.FormConfigurationProvider;
+
+@Model(
+    adaptables = { Resource.class },
+    adapters = FormConfigurationProvider.class)
+public class FormConfigurationProviderImpl implements FormConfigurationProvider {
+
+    private static final String CUSTOM_FUNCTION_CONFIG_BUCKET_NAME = "settings/cloudconfigs";
+    private static final String CUSTOM_FUNCTION_CONFIG_NAME = "edge-delivery-service-configuration";
```
Comment (rismehta) [inline bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/form/FormConfigurationProviderImpl.java:37]: Isn't this borrowed from SITES, we should re-use the constants and not define our own

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/form/FormConfigurationProviderImpl.java (hunk 3 above)
Comment (jalagari) [inline bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/form/FormConfigurationProviderImpl.java:59]: @deepprakash345 - Edge Delivery Configuration now have option to select the page/live, please reuse it instead of fixed hlx.live

File: ui.tests/test-module/specs/customfunctionedgedelivery/customfunctionedgedelivery.spec.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                toggle_array = response.body.enabled;
+            }
+        });
+    });
+
+
+
+    it(" fullName field should update on change in first name and last name ", () => {
+        if(toggle_array.includes("FT_FORMS-11269") && toggle_array.includes("FT_FORMS-12423")) {
+            const firstName = Object.entries(formContainer._fields)[0];
```
Comment (rismehta) [inline ui.tests/test-module/specs/customfunctionedgedelivery/customfunctionedgedelivery.spec.js:40]: Ideally test should only depend on FT, if the PR change list has the FT in the toggle provider JSON ? Please remove the FT from these test cases

File: ui.frontend/src/utils.js [hunk 6]
```diff
@@ -331,6 +332,22 @@ class Utils {
         }
     }
 
+    static async registerCustomFunctionsByUrl(url) {
+        if (url != null && url.trim().length > 0) {
+            // webpack ignore is added because webpack was converting this to a static import upon bundling resulting in error.
+            const customFunctionModule = await import(/*webpackIgnore: true*/ url);
```
Comment (rismehta) [inline ui.frontend/src/utils.js:338]: Can you add a comment here that the URL should have the CORS whitelist configured

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/container.html [hunk 7]
```diff
@@ -38,6 +41,7 @@
       data-cmp-page-lang="${container.containingPageLang}"
       data-cmp-path="${resource.path}"
       class="cmp-adaptiveform-container cmp-container ${wcmmode.edit ? 'cmp-adaptiveform-container--edit' : ''}"
+      data-cmp-custom-functions-module-url="${configurationProvider.customFunctionModuleUrl}"
```
Comment (rismehta) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/container.html:44]: Please update the component read me about this new data attribute and its purpose

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormConfigurationProvider.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+
+package com.adobe.cq.forms.core.components.models.form;
+
+import org.osgi.annotation.versioning.ProviderType;
+
+@ProviderType
+public interface FormConfigurationProvider {
```
Comment (rismehta) [inline bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormConfigurationProvider.java:22]: Please bump the minor version of package-info and update the java doc using @since annotation

Comment (rismehta) [inline bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormConfigurationProvider.java:22]: Also add java doc for the interface

**Review by rismehta — CHANGES_REQUESTED** [review]
check comments

**Comment by rismehta** [comment]
the patch coverage is less, please fix the coverage

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/form/FormConfigurationProviderImpl.java (hunk 3 above)
Comment (deepprakash345) [reply bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/form/FormConfigurationProviderImpl.java:59]: Hi Vijay,
This interface  from sites franklin does not expose any such api. Once this PR 
https://git.corp.adobe.com/livecycle/cq-guides/pull/6156
is merged in cq-guides and released in far, we will be using the same api to fetch this details.

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/form/FormConfigurationProviderImpl.java (hunk 4 above)
Comment (deepprakash345) [reply bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/form/FormConfigurationProviderImpl.java:37]: Hi Rishi,
Once this PR
https://git.corp.adobe.com/livecycle/cq-guides/pull/6156
is merged in cq-guides and released in far, we will be using the same api to fetch the details rather than writing this logic here. These constants are not exposed by sites-franklin

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormConfigurationProvider.java (hunk 8 above)
Comment (deepprakash345) [reply bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormConfigurationProvider.java:26]: done

Comment (deepprakash345) [reply bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormConfigurationProvider.java:26]: done

File: ui.frontend/src/utils.js (hunk 6 above)
Comment (deepprakash345) [reply ui.frontend/src/utils.js:339]: done

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/container.html (hunk 2 above)
Comment (deepprakash345) [reply ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/container.html]: done

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/form/FormConfigurationProviderImpl.java (hunk 3 above)
Comment (deepprakash345) [reply bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/form/FormConfigurationProviderImpl.java]: done

File: ui.tests/test-module/specs/customfunctionedgedelivery/customfunctionedgedelivery.spec.js (hunk 5 above)
Comment (deepprakash345) [reply ui.tests/test-module/specs/customfunctionedgedelivery/customfunctionedgedelivery.spec.js:40]: added FT's in toggle provider JSON

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/container.html (hunk 7 above)
Comment (deepprakash345) [reply ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/container.html:43]: done

**Comment by arunattri26** [comment]
@deepprakash345 , is it possible to move all custom function edge delivery collateral and tests in a folder under 'ruleeditor' folder?
We are planning to run rule editor tests in rule editor pipelines as well and would like to keep all rule editor specific tests under single folder.

**Review by arunattri26 — DISMISSED** [review]
@deepprakash345 , is it possible to move all custom function edge delivery collateral and tests in a folder under 'ruleeditor' folder?
We are planning to run rule editor tests in rule editor pipelines as well and would like to keep all rule editor specific tests under single folder.

**Comment by deepprakash345** [comment]
> @deepprakash345 , is it possible to move all custom function edge delivery collateral and tests in a folder under 'ruleeditor' folder? We are planning to run rule editor tests in rule editor pipelines as well and would like to keep all rule editor specific tests under single folder.

@arunattri26 , the custom function tests are of runtime, lets discuss this over a call, we are not testing any rule editor releated functionality in these tests

**Comment by deepprakash345** [comment]
> the patch coverage is less, please fix the coverage

Screenshot 2024-05-21 at 9 40 47 AM

@rismehta , as seen in the screenshot of the codecov, all the branches of the code are covered, its showing the **ifs** as partial lines.

File: ui.tests/test-module/specs/customfunctionedgedelivery/customfunctionedgedelivery.spec.js (hunk 5 above)
Comment (rismehta) [reply ui.tests/test-module/specs/customfunctionedgedelivery/customfunctionedgedelivery.spec.js:40]: Are these FT's dependent on the feature which is part of this PR ? If not, please remove these from the CL

**Comment by jalagari** [comment]
> > the patch coverage is less, please fix the coverage
> 
> Screenshot 2024-05-21 at 9 40 47 AM
> @rismehta , as seen in the screenshot of the codecov, all the branches of the code are covered, its showing the **ifs** as partial lines.

@deepprakash345 - Can we add unit test cases for missing cases?

**Review by rismehta — CHANGES_REQUESTED** [review]
Are the FT present in the test case, dependent on the feature which is part of this PR ? If not, please remove these from the CL

**Comment by deepprakash345** [comment]
> Are the FT present in the test case, dependent on the feature which is part of this PR ? If not, please remove these from the CL

removed

**Comment by deepprakash345** [comment]
> > > the patch coverage is less, please fix the coverage
> > 
> > 
> > Screenshot 2024-05-21 at 9 40 47 AM
> > @rismehta , as seen in the screenshot of the codecov, all the branches of the code are covered, its showing the **ifs** as partial lines.
> 
> @deepprakash345 - Can we add unit test cases for missing cases?

added UTs , coverage check passed.

**Comment by deepprakash345** [comment]
@rismehta , @jalagari please check

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/_cq_dialog/.content.xml (hunk 1 above)
Comment (deepprakash345) [reply ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/_cq_dialog/.content.xml:74]: added spec Version

File: ui.tests/test-module/specs/customfunctionedgedelivery/customfunctionedgedelivery.spec.js (hunk 5 above)
Comment (deepprakash345) [reply ui.tests/test-module/specs/customfunctionedgedelivery/customfunctionedgedelivery.spec.js:39]: done

**Review by rismehta — APPROVED** [review]
